src/qmapshack/poi/CPoiGPX.cpp

```diff
@@ -0,0 +1,379 @@
+/**********************************************************************************************
+    Copyright (C) 2020 Oliver Eichler <oliver.eichler@gmx.de>
+    Copyright (C) 2021 Henri Hornburg <pingurus@t-online.de>
+
+    This program is free software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+**********************************************************************************************/
+
+#include "gis/Poi.h"
+#include "gis/wpt/CGisItemWpt.h"
+#include "helpers/CDraw.h"
+#include "poi/CPoiCategory.h"
+#include "poi/CPoiDraw.h"
+#include "poi/CPoiGPX.h"
+#include "poi/CPoiIconCategory.h"
+#include "poi/IPoi.h"
+
+#include <QtWidgets>
+
+CPoiGPX::CPoiGPX(const QString &filename, CPoiDraw *parent)
+    : IPoi(parent)
+    , filename(filename)
+    , loadTimer(new QTimer(this))
+{
+    //Set up a timer to only start reloading and redrawing the map when the user has made his selections
+    loadTimer->setSingleShot(true);
+    loadTimer->setInterval(500);
+    connect(loadTimer, &QTimer::timeout, poi, &CPoiDraw::emitSigCanvasUpdate);
+
+    //Read waypoints from file
+    // create file instance
+    QFile file(filename);
+
+    // load file content to xml document
+    QDomDocument xml;
+    QString msg;
+    int line;
+    int column;
+    if(!xml.setContent(&file, false, &msg, &line, &column))
+    {
+        file.close();
+        throw tr("Failed to read: %1\nline %2, column %3:\n %4").arg(filename).arg(line).arg(column).arg(msg);
+    }
+    file.close();
+
+    QDomElement xmlGpx = xml.documentElement();
+    if(xmlGpx.tagName() != "gpx")
+    {
+        throw tr("Not a GPX file: %1").arg(filename);
+    }
+
+    const QDomNodeList& xmlWpts = xmlGpx.elementsByTagName("wpt");
+    for(int n = 0; n < xmlWpts.count(); ++n)
+    {
+        const QDomNode& xmlWpt = xmlWpts.item(n);
+        QSharedPointer<CGisItemWpt> wpt = QSharedPointer<CGisItemWpt>(new CGisItemWpt(xmlWpt, nullptr));
```

Comment:
@kiozen is this too hacky? I also added a shared pointer to poi_t, so if that is not a `nullptr`, we can simply copy the `CGisItemWpt` to any project the user whishes.

Reply:
Not sure why we need a shared pointer here. Doesn't  the object load all POIs at the beginning? And wouldn't referencing the POIs in that list be sufficient? After all this list does not change when loaded, right?

And yes I know, `CGisItemWpt` is violating the model/view  architecture and this backlashing, now. But creating literally thousands of parentless QTreeWidgetItems with the additional overhead of CGisItemWpt doesn't feel right.

On the other hand side there is `IGisItem::wpt_t` and `IGisItem::readWpt()` probably could be a static method (haven't tried it yet). Maybe that is a much better approach.

Reply:
I took a shared pointer, as I was not sure about the possibly dangling pointers when the user deactivates the gpx-poi file whilst having a "Select Items on Map" open. It might not end well when the file is closed and the wpts are destroyed and then the user tries to copy them over to a project.

Regarding `IGisItem::wpt_t`: it does not include the geocache information, so I'd prefer to avoid using that. :smile:

Reply:
Pointer: hm, ok, that might be a problem. However I would assume all objects on the map are destroyed when deactivating the file. And the "Select Items on Map"  should be updated in that case. At least for stuff on the workspace it does (partially, the summary is not updated :( ).

And there is `CGisItemWpt::geocache_t` that can be used outside of CGisItemWpt.